### ISSUE TITLE
[manila-csi-plugin] Support for `cloud-config` secret

### ISSUE THREAD

**Issue by GlassOfWhiskey** [kubernetes/cloud-provider-openstack#2532]
/kind feature

**What happened**: All the other OpenStack-related plugins support a `cloud-config` secret that contains credentials for authentication with the keystone. Conversely, Manila CSI Plugin wants its own format for auth secrets, making it difficult to integrate it with existing Kubernetes-on-Openstack environments (e.g., the Charmed Kubernetes Distribution.

**What you expected to happen**: it would be easier to mount the same `cloud-config` secret in all the OpenStack plugins ecosystem, instead of having a different integration path just for Manila. Is it something feasible?

**Comment by jichenjc** [comment]
I don't know detail of Manila about its secret mgmt if what you said apply to it I agree it's reasonable to update 
and Manila used to be done by @gman0 before  @GlassOfWhiskey do you want to work on this by submit PR ?

**Comment by mkjpryor** [comment]
+1 for this.

When we deploy Kubernetes, we create one secret containing a `clouds.yaml` with an application credential inside, then mount that into the Cinder CSI and OCCM pods. I was hoping to do the same for Manila, but this has scuppered me.

It would be less of a problem if Manila didn't require the region to be specified, which should be implicit in the application credential (it is for the other components), although still irritating.

**Comment by mdbooth** [comment]
I'm also interested in this feature. Without having looked, I doubt Manila genuinely needs Region, btw.

**Comment by kayrus** [comment]
@GlassOfWhiskey at first glance it looks like manila has its own specific set of secrets because csi-controller dynamically passes them to the node controller using CSI spec.

**Comment by musaprg** [comment]
Hi. I was wondering if we could make this happen. +1 to the idea, and I'd like to work on it. I'll try to submit a PR for the further discussion.

/reopen
/assign

**Comment by kayrus** [comment]
the support was added in #2883

**Comment by musaprg** [comment]
~~tracked and resolved by https://github.com/kubernetes/cloud-provider-openstack/issues/2886~~

**Comment by musaprg** [comment]
@kayrus Hi. I checked the PR but it's more like supporting `clouds.yaml` and not focusing on supporting `cloud-config` secret. If I understood correctly, this issue aims to support [`cloud-config` secret](https://github.com/kubernetes/cloud-provider-openstack/blob/31fbaa4f17b145b3461fb94a12153ebbbe9ae992/charts/openstack-cloud-controller-manager/README.md#using-an-external-secret) used by other components, such as `occm`, `cinder-csi-plugin` and `barbican-kms-plugin`. I think we should consider supporting it separately from #2883, #2886.

**Comment by kayrus** [comment]
@musaprg do you mean to adjust helm-chart?

**Comment by musaprg** [comment]
@kayrus No, I meant to allow manila-csi-plugin to accept `cloud.conf` other than configurations written directly into Secret as individual fields. Here's the difference between other components and manila-csi-plugin in terms of Secret format. If manila-csi-plugin allows to accept the `cloud.conf`, we could unify the credential into single `cloud-config` Secret for both cinder-csi-plugin and manila-csi-plugin.

## `cloud-config` secret

```
apiVersion: v1
kind: Secret
metadata:
  name: cloud-config
  namespace: kube-system
type: Opaque
stringData:
  cloud.conf: |-
    [Global]
    auth-url="some-auth-url"
    username="some-username"
    password="some-password"
    region="some-region"
    tenant-id="some-project-id"
    domain-name="some-domain-id"
```

## `csi-manila-secrets` secret

```
apiVersion: v1
kind: Secret
metadata:
  name: csi-manila-secrets
  namespace: default
stringData:
  os-authURL: "some-auth-url"
  os-region: "some-region"
  os-userName: "some-username"
  os-password: "some-password"
  os-projectID: "some-project-id"
  os-domainID: "some-domain-id"
```

**Comment by stephenfin** [comment]
As @musaprg notes, https://github.com/kubernetes/cloud-provider-openstack/pull/2883 is a different thing. That `clouds.yaml` support would be both compatible with the existing secret approach and a potential `cloud.conf` file.

This being the case, I'd like to discuss whether we want to do this versus changing how _cinder_ works. afaict, the approach Manila is taking to handle credentials is the "correct" way to handle credentials, as documented in the CSI Developer Docs:

https://kubernetes-csi.github.io/docs/secrets-and-credentials-volume-snapshot-class.html#volumesnapshotclass-secrets

@musaprg given the above, is there a particular reason you'd like to pursue the `cloud.conf` config file approach for Manila other than consistency? If it's purely consistently, then rather than extending Manila to use `cloud.conf`, could we extend Cinder to use secrets and the `csi.storage.k8s.io/snapshotter-secret-name` and `csi.storage.k8s.io/snapshotter-secret-namespace` parameters?

**Comment by musaprg** [comment]
@stephenfin Hi! Thanks for your comment.

> given the above, is there a particular reason you'd like to pursue the cloud.conf config file approach for Manila other than consistency?

No, my suggestion is purely for consistency in configuration format to share single secret for CSI across multiple plugins.

> could we extend Cinder to use secrets and the csi.storage.k8s.io/snapshotter-secret-name and csi.storage.k8s.io/snapshotter-secret-namespace parameters?

That makes sense to make Cinder side compliant with the CSI-way, but seems slightly different topic.

The original issue complains that only Manila CSI plugins accept credentials in completely different format, which makes it hard to use single secret across all plugins hosted in cloud-provider-openstack. If we decide to switch to individual keys in secret rather than `cloud.conf`, other plugins should have to support it.

We have two topics to be discussed here.
1. (How to pass credentials) Support accepting credentials via Secret in other OpenStack plugins
2. (Credential format) Support individual keys format rather than `cloud.conf` in all OpenStack plugins

IMO, supporting cloud.conf format in manila plugin would introduce less burden to achieve the goal of this issue. Shifting to individual keys could be proceeded independently, which is out of scope for this issue.

What are your thoughts on that?

**Comment by musaprg** [comment]
> If it's purely consistently, then rather than extending Manila to use cloud.conf, could we extend Cinder to use secrets and the csi.storage.k8s.io/snapshotter-secret-name and csi.storage.k8s.io/snapshotter-secret-namespace parameters?

Now I realized that my original motivation (sharing same secret across multiple CSI plugins) can be achieved in a way suggested by @stephenfin. I've created another issue #2963 to discuss the support of credentials paassed via CSI requests on the Cinder CSI plugin. Can we discuss there?

https://github.com/kubernetes/cloud-provider-openstack/issues/2963

I'll put my work for this issue on hold until the discussion of #2963 is concluded.
